shim.go

```diff
@@ -0,0 +1,113 @@
+// Copyright 2017 HyperHQ Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+package main
+
+import (
+	"fmt"
+	"io"
+	"os"
+	"os/signal"
+	"sync"
+	"syscall"
+
+	"github.com/golang/glog"
+	"github.com/moby/moby/pkg/term"
+)
+
+type Shim struct {
+	container string
+	process   string
+	agent     shimAgent
+}
+
+func NewShim(container, process, addr string) (*Shim, error) {
+	if agent, err := NewShimAgent(addr); err != nil {
+		return nil, err
+	} else {
+		return &Shim{container: container,
+			process: process,
+			agent:   agent}, nil
+	}
+}
+
+func (s *Shim) proxyStdio(wg *sync.WaitGroup) {
+	// don't wait the copying of the stdin, because `io.Copy(inPipe, os.Stdin)`
+	// can't terminate when no input. todo: find a better way.
+	wg.Add(2)
+	inPipe, outPipe, errPipe := shimStdioPipe(s.agent, s.container, s.process)
+	go func() {
+		_, err1 := io.Copy(inPipe, os.Stdin)
+		err2 := s.agent.CloseStdin(s.container, s.process)
+		glog.V(3).Infof("copy stdin err1 %s err2 %s", err1, err2)
+	}()
+
+	go func() {
+		_, err := io.Copy(os.Stdout, outPipe)
+		glog.V(3).Infof("copy stdout %s", err)
+		wg.Done()
+	}()
+
+	go func() {
+		_, err := io.Copy(os.Stderr, errPipe)
+		glog.V(3).Infof("copy stderr %s", err)
+		wg.Done()
+	}()
+}
+
+func (s *Shim) forwardAllSignals() chan os.Signal {
+	sigc := make(chan os.Signal, 2048)
+	// handle all signals for the process.
+	signal.Notify(sigc)
+	signal.Ignore(syscall.SIGCHLD, syscall.SIGPIPE)
+
+	go func() {
+		for sig := range sigc {
+			if sig == syscall.SIGCHLD || sig == syscall.SIGPIPE || sig == syscall.SIGWINCH {
```

Comment:
We need to consider a few more signals here like SIGBUS, SIGSEGV and SIGABRT. These should be ignored as well as they may indicate runtime exceptions with the shim itself.